src/main/java/com/cloudant/client/org/lightcouch/CouchDatabaseBase.java

```diff
@@ -338,7 +334,9 @@ public Response saveAttachment(InputStream in, String name, String contentType,
         assertNotEmpty(in, "in");
         assertNotEmpty(name, "name");
         assertNotEmpty(contentType, "ContentType");
-        assertNotEmpty(docId, "docId");
+        if (docId == null) {
+            docId = generateUUID();
+        }
```

Comment:
Thanks Al, that is a good spot.
I've re-added the previous behaviour for a non-null docId, i.e. verifying it is not an empty string.
Additionally I added checks to ensure that if docId has been provided that the revision is either `null` or a non-empty string (for the create and update cases respectively).
All in 3b75b3d including 3 new tests for these `IllegalArgumentException` cases.